<a href="https://colab.research.google.com/github/Nikita-Gilev/math_stat/blob/lesson_2/%D0%9A%D0%BE%D1%80%D1%80%D0%B5%D0%BA%D1%82%D0%B8%D1%80%D0%BE%D0%B2%D0%BA%D0%B0_%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D1%8F_%D0%BA_%D1%83%D1%80%D0%BE%D0%BA%D1%83_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
import numpy as np
from decimal import Decimal
from scipy.stats import poisson
from scipy.stats import binom

In [5]:
import scipy.stats as sps

In [6]:
import matplotlib.pyplot as plt

In [7]:
from math import *

In [8]:
def combination (m, n):
    c = np.math.factorial(n)//(np.math.factorial(m)*np.math.factorial(n - m))
    return c

In [9]:
def bernulli (k, n, p):
    P = combination(k, n) * p**k * (1-p)**(n-k)
    return P

In [10]:
def p_poisson(k, p_lambda):
    P = (p_lambda**k) * (np.exp(-p_lambda))/np.math.factorial(k)
    return P

Задача 1

Контрольная работа состоит из пяти вопросов. На каждый вопрос приведено четыре варианта ответа, один из которых правильный. Составьте закон распределения числа правильных ответов при простом угадывании. Найдите `M(X)`, `D(X)`.

Выбор правильного ответа случаен для любого из вопросов, вероятность выбора правильного ответа одинакова для всех вопросов и соотетствует 0,25.    
Воспользуюсь формулой Бернулли со следующими параметрами n = 4, p =0.25, k=0...4, где к - количество правильных ответов.



In [11]:
for k in range(5):
    r = bernulli(k, 4, 0.25)
    print(r)

0.31640625
0.421875
0.2109375
0.046875
0.00390625


распределение вероятности выбора правильных ответов x:
<table border="3">
    <tr>
        <td>k</td>
        <td>0</td>
        <td>1</td>
        <td>2</td>
        <td>3</td>
        <td>4</td>
    </tr>
    <tr>
        <td>P(X=k)</td>
        <td>0.31640625</td>
        <td>0.421875</td>
        <td>0.2109375</td>
        <td>0.046875</td>
        <td>0.00390625</td>
    </tr>
</table>

Математическое ожидание для биноминалього распределения:  
M(X) = np

In [12]:
math_expect = 4 * 0.25
print(math_expect)

1.0


Дисперсия для биноминального распределения:

In [13]:
D = 4 * 0.25 * (1-0.25)
print(D)

0.75


Задача 2

Вероятность поймать рыбу при одном забрасывании удочки равна 0.01. Какова вероятность того, что число пойманных рыб при 200 забросах удочки будет не менее 5 и не более 10?

Применим распределение Пуассона.  
$λ = np$.  
p = 0.25  
n = 200  
5 <= k <= 10

In [14]:
p_lambda = 200 * 0.01
print(p_lambda)

2.0


In [15]:
total = 0
for i in range (5, 11):
    probability = p_poisson(i, p_lambda)
    print(probability)
    total += probability

0.03608940886309672
0.012029802954365574
0.0034370865583901638
0.0008592716395975409
0.00019094925324389798
3.8189850648779595e-05


In [19]:
print(f"Вероятность выловить от 5 до 10 (включительно) рыб при данных параметрах {round(total, 6) * 100} ")

Вероятность выловить от 5 до 10 (включительно) рыб при данных параметрах 5.2645 


**Задача 3**

В условиях предыдущей задачи сколько забросов удочки нужно сделать чтобы быть уверенным с вероятностью 0.9, что хотя бы 10 раз будет поймана рыба?

_Подсказка_. Как посчитать `P(X >= k)` для величины `X`, имеющей биномиальное распределение? Здесь предстоит немножко покодить.


Суммарная вероятность = 0.9, i = 10

P(X=n) = 0.9  
n=?

In [ ]:
p = 0.01
k = 10

P(X >= 10) = 1 - P(X < 10)

Применим метод Пуассона

In [18]:
for n in range (300, 5000):
    p_lambda = 0.01 * n
    P = 1 - sum(p_poisson(i, p_lambda) for i in range(10))    
    if P > 0.099 and P < 0.1:        
        print(f"Полученное значение вероятности {P} достигнуто при {n} попытках")
    if P > 0.1:        
        print(f"Полученное значение вероятности {P} достигнуто при {n} попытках")
        break
        


Полученное значение вероятности 0.09913811522524696 достигнуто при 621 попытках
Полученное значение вероятности 0.09990031163830515 достигнуто при 622 попытках
Полученное значение вероятности 0.10066592227501103 достигнуто при 623 попытках


In [17]:
for n in range (300, 5000):
    p_lambda = 0.01 * n
    P = 1 - sum(p_poisson(i, p_lambda) for i in range(10))    
    if P > 0.899 and P < 0.9:        
        print(f"Полученное значение вероятности {P} достигнуто при {n} попытках")
    if P > 0.9:        
        print(f"Полученное значение вероятности {P} достигнуто при {n} попытках")
        break

Полученное значение вероятности 0.89929526467099 достигнуто при 1419 попытках
Полученное значение вероятности 0.8997364748565644 достигнуто при 1420 попытках
Полученное значение вероятности 0.9001760722984115 достигнуто при 1421 попытках
